ui_test/src/lib.rs

```diff
@@ -0,0 +1,414 @@
+
+                        // Using `format` to prevent messages from threads from getting intermingled.
+                        let mut msg = format!("{} ", path.display());
+                        if !revision.is_empty() {
+                            msg = format!("{msg}(revision `{revision}`) ");
+                        }
+                        msg = format!("{msg} .. ");
+                        if errors.is_empty() {
+                            eprintln!("{msg}{}", "ok".green());
+                        } else {
+                            eprintln!("{msg}{}", "FAILED".red().bold());
+                            failures.lock().unwrap().push((path.clone(), m, revision, errors));
+                        }
+                    }
+                }
+            });
+        }
+    })
+    .unwrap();
+
+    // Print all errors in a single thread to show reliable output
+    let failures = failures.into_inner().unwrap();
+    let total = total.load(Ordering::Relaxed);
+    let skipped = skipped.load(Ordering::Relaxed);
+    if !failures.is_empty() {
+        for (path, miri, revision, errors) in &failures {
+            eprintln!();
+            eprint!("{}", path.display().to_string().underline());
+            if !revision.is_empty() {
+                eprint!(" (revision `{}`)", revision);
+            }
+            eprint!("{}", " FAILED".red());
+            eprintln!();
+            eprintln!("command: {:?}", miri);
+            eprintln!();
+            let mut dump_stderr = None;
+            for error in errors {
+                match error {
+                    Error::ExitStatus(mode, exit_status) => eprintln!("{mode:?} got {exit_status}"),
+                    Error::PatternNotFound { stderr, pattern, definition_line } => {
+                        eprintln!("`{pattern}` {} in stderr output", "not found".red());
+                        eprintln!(
+                            "expected because of pattern here: {}:{definition_line}",
+                            path.display()
+                        );
+                        dump_stderr = Some(stderr.clone())
+                    }
+                    Error::NoPatternsFound =>
+                        eprintln!("{}", "no error patterns found in failure test".red()),
+                    Error::PatternFoundInPassTest =>
+                        eprintln!("{}", "error pattern found in success test".red()),
+                    Error::OutputDiffers { path, actual, expected } => {
+                        dump_stderr = None;
+                        compare_output(path, actual, expected);
+                    }
+                }
+                eprintln!();
+            }
+            if let Some(stderr) = dump_stderr {
+                eprintln!("actual stderr:");
+                eprintln!("{}", stderr);
+                eprintln!();
+            }
+        }
+        eprintln!(
+            "{} tests failed, {} tests passed, {} skipped",
```

Comment:
This can be potentially confusing since for "total" and "skipped" we count a test with many revisions as 1, but for failures we count it as multiple tests.